File: Sources/HolodeckCore/Models/Appearance.swift
// MIT License
//
// Copyright (c) 2026 Otávio C.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import Foundation

public enum Appearance: String, Sendable, CaseIterable {

    case light
    case dark
}

public enum BatteryState: String, Sendable, CaseIterable {

    case charging
    case charged
    case discharging
}

public struct StatusBarOverrides: Sendable, Equatable {

    // MARK: - Properties

    public var time: String?
    public var batteryState: BatteryState?
    public var batteryLevel: Int?
    public var wifiBars: Int?
    public var cellularBars: Int?
    public var operatorName: String?

    // MARK: - Lifecycle

    public init(
        time: String? = nil,
        batteryState: BatteryState? = nil,
        batteryLevel: Int? = nil,
        wifiBars: Int? = nil,
        cellularBars: Int? = nil,
        operatorName: String? = nil
    ) {
        self.time = time
        self.batteryState = batteryState
        self.batteryLevel = batteryLevel
        self.wifiBars = wifiBars
        self.cellularBars = cellularBars
        self.operatorName = operatorName
    }

    // MARK: - Public

    public var isEmpty: Bool {
        time == nil
            && batteryState == nil
            && batteryLevel == nil
            && wifiBars == nil
            && cellularBars == nil
            && operatorName == nil
    }

    public var simctlArguments: [String] {
        var args: [String] = []
        if let time { args += ["--time", time] }
        if let batteryState { args += ["--batteryState", batteryState.rawValue] }
        if let batteryLevel { args += ["--batteryLevel", String(batteryLevel)] }
        if let wifiBars { args += ["--wifiBars", String(wifiBars)] }
        if let cellularBars { args += ["--cellularBars", String(cellularBars)] }
        if let operatorName { args += ["--operatorName", operatorName] }
        return args
    }
}

File: crates/holodeck-core/src/models/appearance.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Appearance {
    Light,
    Dark,
}

impl Appearance {
    pub fn raw_value(self) -> &'static str {
        match self {
            Appearance::Light => "light",
            Appearance::Dark => "dark",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BatteryState {
    Charging,
    Charged,
    Discharging,
}

impl BatteryState {
    pub fn raw_value(self) -> &'static str {
        match self {
            BatteryState::Charging => "charging",
            BatteryState::Charged => "charged",
            BatteryState::Discharging => "discharging",
        }
    }
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct StatusBarOverrides {
    pub time: Option<String>,
    pub battery_state: Option<BatteryState>,
    pub battery_level: Option<i64>,
    pub wifi_bars: Option<i64>,
    pub cellular_bars: Option<i64>,
    pub operator_name: Option<String>,
}

impl StatusBarOverrides {
    pub fn is_empty(&self) -> bool {
        self.time.is_none()
            && self.battery_state.is_none()
            && self.battery_level.is_none()
            && self.wifi_bars.is_none()
            && self.cellular_bars.is_none()
            && self.operator_name.is_none()
    }

    pub fn simctl_arguments(&self) -> Vec<String> {
        let mut args = Vec::new();
        if let Some(time) = &self.time {
            args.push("--time".to_string());
            args.push(time.clone());
        }
        if let Some(state) = self.battery_state {
            args.push("--batteryState".to_string());
            args.push(state.raw_value().to_string());
        }
        if let Some(level) = self.battery_level {
            args.push("--batteryLevel".to_string());
            args.push(level.to_string());
        }
        if let Some(bars) = self.wifi_bars {
            args.push("--wifiBars".to_string());
            args.push(bars.to_string());
        }
        if let Some(bars) = self.cellular_bars {
            args.push("--cellularBars".to_string());
            args.push(bars.to_string());
        }
        if let Some(name) = &self.operator_name {
            args.push("--operatorName".to_string());
            args.push(name.clone());
        }
        args
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_overrides_report_empty() {
        assert!(StatusBarOverrides::default().is_empty());
    }

    #[test]
    fn builds_simctl_arguments_in_order() {
        let overrides = StatusBarOverrides {
            time: Some("9:41".to_string()),
            battery_level: Some(100),
            ..Default::default()
        };
        assert_eq!(overrides.simctl_arguments(), vec!["--time", "9:41", "--batteryLevel", "100"]);
    }
}
